// hardhat-uniswap/src/v3/deployers/deployTokenDescriptor.ts
import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
import { abi, bytecode } from "@uniswap/v3-periphery/artifacts/contracts/NonfungibleTokenPositionDescriptor.sol/NonfungibleTokenPositionDescriptor.json";
import { Contract, ContractFactory } from "ethers";
import { linkLibraries } from "../../util/linkLibraries";

async function deployTokenDescriptor(
  signer: SignerWithAddress,
  nftDescriptorLibrary: Contract,
  weth9: Contract, 
) {
  const linkedBytecode = linkLibraries(
    {
      bytecode: bytecode,
      linkReferences: {
        "NFTDescriptor.sol": {
          NFTDescriptor: [
            {
              length: 20,
              start: 1261,
            },
          ],
        },
      },
    },
    {
      NFTDescriptor: nftDescriptorLibrary.target,
    }
  )
  const TokenDescriptor = new ContractFactory(abi, linkedBytecode, signer);
  const tokenDescriptor = await TokenDescriptor.deploy(weth9.target);
  return { tokenDescriptor, TokenDescriptor };
}

export default deployTokenDescriptor;
