// hardhat-uniswap/dist/src/v3/deployers/deployTokenDescriptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const NonfungibleTokenPositionDescriptor_json_1 = require("@uniswap/v3-periphery/artifacts/contracts/NonfungibleTokenPositionDescriptor.sol/NonfungibleTokenPositionDescriptor.json");
const ethers_1 = require("ethers");
const linkLibraries_1 = require("../../util/linkLibraries");
async function deployTokenDescriptor(signer, nftDescriptorLibrary, weth9) {
    const linkedBytecode = (0, linkLibraries_1.linkLibraries)({
        bytecode: NonfungibleTokenPositionDescriptor_json_1.bytecode,
        linkReferences: {
            "NFTDescriptor.sol": {
                NFTDescriptor: [
                    {
                        length: 20,
                        start: 1261,
                    },
                ],
            },
        },
    }, {
        NFTDescriptor: nftDescriptorLibrary.target,
    });
    const TokenDescriptor = new ethers_1.ContractFactory(NonfungibleTokenPositionDescriptor_json_1.abi, linkedBytecode, signer);
    const tokenDescriptor = await TokenDescriptor.deploy(weth9.target);
    return { tokenDescriptor, TokenDescriptor };
}
exports.default = deployTokenDescriptor;
